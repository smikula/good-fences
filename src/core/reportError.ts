import * as path from 'path';
import getOptions from '../utils/getOptions';
import Config from '../types/config/Config';

export default function reportError(
    message: string,
    sourceFile: string,
    rawImport: string,
    config: Config
) {
    let fencePath = config.path + path.sep + 'fence.json';

    let detailedMessage =
        `Good-fences violation in ${sourceFile}:\n` +
        `    ${message}: ${rawImport}\n` +
        `    Fence: ${fencePath}`;

    if (getOptions().onError) {
        getOptions().onError({
            message,
            sourceFile,
            rawImport,
            fencePath,
            detailedMessage,
        });
    }
}
