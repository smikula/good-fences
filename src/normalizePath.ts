import * as path from 'path';
import NormalizedPath from './types/NormalizedPath';

export default function normalizePath(...pathSegments: string[]) {
    // Resolve the raw path to an absolute path
    let normalizedPath = path.resolve.apply(null, pathSegments);

    // Normalize drive letters to upper case
    if (normalizedPath.match(/^[a-z]:/)) {
        normalizedPath = normalizedPath.substr(0, 1).toUpperCase() + normalizedPath.substr(1);
    }

    return <NormalizedPath>normalizedPath;
}
