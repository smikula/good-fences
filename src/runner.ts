import Options from './types/Options';
import getOptions, { setOptions } from './getOptions';
import validateFile from './validateFile';
import TypeScriptProgram from './TypeScriptProgram';
import createPath from './createPath';

export function run(options: Options) {
    // Store options so they can be globally available
    setOptions(options);

    // Run validation
    let tsProgram = new TypeScriptProgram(getOptions().project);
    let files = tsProgram.getSourceFiles();
    files.forEach(file => {
        validateFile(createPath(file), tsProgram);
    });
}
