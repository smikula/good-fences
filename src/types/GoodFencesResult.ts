import ValidationError from './ValidationError';

export default interface GoodFencesResult {
    errors: ValidationError[];
    warnings: ValidationError[];
};
