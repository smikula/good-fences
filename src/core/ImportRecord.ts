import NormalizedPath from '../types/NormalizedPath';
import TypeScriptProgram from './TypeScriptProgram';
import normalizePath from '../utils/normalizePath';

export default class ImportRecord {
    public filePath: NormalizedPath;

    constructor(importSource: string, sourceFile: NormalizedPath, tsProgram: TypeScriptProgram) {
        const resolvedFileName = tsProgram.resolveImportFromFile(importSource, sourceFile);
        if (resolvedFileName) {
            this.filePath = normalizePath(resolvedFileName);
        }
    }
}
