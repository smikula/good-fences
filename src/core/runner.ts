import RawOptions from '../types/RawOptions';
import getOptions, { setOptions } from '../utils/getOptions';
import validateFile from '../validation/validateFile';
import TypeScriptProgram from './TypeScriptProgram';
import normalizePath from '../utils/normalizePath';

export function run(rawOptions: RawOptions) {
    // Store options so they can be globally available
    setOptions(rawOptions);

    // Run validation
    let tsProgram = new TypeScriptProgram(getOptions().project);
    let files = tsProgram.getSourceFiles();
    files.forEach(file => {
        validateFile(normalizePath(file), tsProgram);
    });
}
