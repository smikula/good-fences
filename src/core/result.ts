import * as path from 'path';
import Config from '../types/config/Config';
import GoodFencesError from '../types/GoodFencesError';
import GoodFencesResult from '../types/GoodFencesResult';
import ImportRecord from './ImportRecord';

const result: GoodFencesResult = {
    errors: [],
    warnings: [],
};

export function getResult() {
    return result;
}

export function reportViolation(
    message: string,
    sourceFile: string,
    importRecord: ImportRecord,
    config: Config
) {
    let fencePath = config.path + path.sep + 'fence.json';

    let detailedMessage =
        `Good-fences violation in ${sourceFile}:\n` +
        `    ${message}: ${importRecord.rawImport}\n` +
        `    Fence: ${fencePath}`;

    const error: GoodFencesError = {
        message,
        sourceFile,
        rawImport: importRecord.rawImport,
        fencePath,
        detailedMessage,
    };

    result.errors.push(error);
}

export function reportConfigError(message: string, config: Config) {
    let fencePath = config.path + path.sep + 'fence.json';

    let detailedMessage =
        `Good-fences configuration error: ${message}\n` + `    Fence: ${fencePath}`;

    const error: GoodFencesError = {
        message,
        fencePath,
        detailedMessage,
    };

    result.errors.push(error);
}

export function reportWarning(message: string, config: Config) {
    let fencePath = config.path + path.sep + 'fence.json';
    let detailedMessage = `Good-fences warning: ${message}\n` + `    Fence: ${fencePath}`;

    const warning: GoodFencesError = {
        message,
        fencePath,
        detailedMessage,
    };

    result.warnings.push(warning);
}
