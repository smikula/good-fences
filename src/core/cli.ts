import * as commander from 'commander';
import Options from '../types/Options';
import { run } from './runner';

// Read the package version from package.json
const packageVersion = require('../../package').version;

// Parse command line options
const options = commander
    .version(packageVersion)
    .option('-p, --project <string>', 'tsconfig.json file')
    .option('-r, --rootDir <string>', 'root directory of the project')
    .parse(process.argv) as Options;

let hadError = false;

// Run good-fences
run({
    ...options,
    onError(error) {
        console.error(error.detailedMessage);
        hadError = true;
    },
});

// Indicate success or failure via the exit code
process.exitCode = hadError ? 1 : 0;
