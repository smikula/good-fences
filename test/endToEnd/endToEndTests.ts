import { run } from '../../src/core/runner';
import GoodFencesResult from '../../src/types/GoodFencesResult';
import normalizePath from '../../src/utils/normalizePath';

describe('runner', () => {
    it('returns the expected results', () => {
        // Arrange
        const expectedResults = require('./endToEndTests.expected.json');

        // Act
        const actualResults = run({
            rootDir: './sample',
        });

        // Assert
        removeDetailedMessages(actualResults);
        normalizePaths(expectedResults);
        expect(actualResults).toEqual(expectedResults);
    });
});

function removeDetailedMessages(results: GoodFencesResult) {
    for (const error of results.errors) {
        delete error.detailedMessage;
    }

    for (const warning of results.warnings) {
        delete warning.detailedMessage;
    }
}

function normalizePaths(results: GoodFencesResult) {
    for (const error of results.errors) {
        error.fencePath = normalizePath(error.fencePath);
        error.sourceFile = normalizePath(error.sourceFile);
    }

    for (const warning of results.warnings) {
        warning.fencePath = normalizePath(warning.fencePath);
    }
}
