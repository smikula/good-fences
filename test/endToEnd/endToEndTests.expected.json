{
    "errors": [
        {
            "message": "Module is not exported",
            "sourceFile": "sample/src/componentA/helperA1.ts",
            "rawImport": "../componentB/helperB1",
            "fencePath": "sample/src/componentB/fence.json"
        },
        {
            "message": "Import not allowed",
            "sourceFile": "sample/src/componentA/helperA1.ts",
            "rawImport": "../componentB/helperB1",
            "fencePath": "sample/src/componentA/fence.json"
        },
        {
            "message": "Import not allowed",
            "sourceFile": "sample/src/componentA/helperA2.ts",
            "rawImport": "../utils/util1",
            "fencePath": "sample/src/componentA/fence.json"
        },
        {
            "message": "Import not allowed",
            "sourceFile": "sample/src/componentA/componentA.ts",
            "rawImport": "../componentB/componentB",
            "fencePath": "sample/src/componentA/fence.json"
        },
        {
            "message": "Module is not exported",
            "sourceFile": "sample/src/index.ts",
            "rawImport": "./componentB/componentB",
            "fencePath": "sample/src/componentB/fence.json"
        }
    ],
    "warnings": [
        {
            "message": "Tag 'unknownTag' is referred to but is not defined in any fence.",
            "fencePath": "sample/src/componentA/fence.json"
        }
    ]
}